test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -770,6 +770,90 @@ class test_ucp_sockaddr : public ucp_test {
         return false;
     }
 
+    void do_force_close_during_rndv(bool fail_send_ep)
+    {
+        constexpr size_t length = 4 * UCS_KBYTE;
+
+        listen_and_communicate(false, SEND_DIRECTION_BIDI);
+
+        mem_buffer recv_buffer(length, UCS_MEMORY_TYPE_HOST);
+        recv_buffer.pattern_fill(2, length);
+        void *rreq = recv(receiver(), recv_buffer.ptr(), length,
+                          rtag_complete_check_data_cbx,
+                          reinterpret_cast<void*>(&recv_buffer));
+
+        mem_buffer send_buffer(length, UCS_MEMORY_TYPE_HOST);
+        send_buffer.pattern_fill(1, length);
+        void *sreq = send(sender(), send_buffer.ptr(), length,
+                          send_recv_type(), scomplete_reset_data_cbx,
+                          reinterpret_cast<void*>(&send_buffer));
+
+        ucp_ep_h ep = sender().revoke_ep();
+
+        // Wait for the TAG RNDV/RTS packet sent and the request scheduled to
+        // be tracked until RNDV/ATS packet is not received from a peer
+        ucs_time_t deadline = ucs::get_deadline();
+        do {
+            ASSERT_LT(ucs_get_time(), deadline);
+            ucp_worker_progress(sender().worker());
+        } while (ucs_hlist_is_empty(&ucp_ep_ext_gen(ep)->proto_reqs));
+        EXPECT_FALSE(ucs_hlist_is_empty(&ucp_ep_ext_gen(ep)->proto_reqs));
+
+        std::map<uct_iface_h, uct_iface_ops_t> sender_uct_ops;
+
+        // Prevent destroying UCT endpoints from discarding to not detect error
+        // by the receiver earlier than data could invalidate by the sender
+        for (auto lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+            if (lane == ucp_ep_get_cm_lane(ep)) {
+                continue;
+            }
+
+            uct_iface_h uct_iface = ep->uct_eps[lane]->iface;
+            auto res = sender_uct_ops.emplace(uct_iface, uct_iface->ops);
+            if (res.second) {
+                uct_iface->ops.ep_flush =
+                        reinterpret_cast<uct_ep_flush_func_t>(
+                                ucs_empty_function_return_no_resource);
+                uct_iface->ops.ep_pending_add =
+                        reinterpret_cast<uct_ep_pending_add_func_t>(
+                                ucs_empty_function_return_busy);
+            }
+        }
+
+        if (fail_send_ep) {
+            UCS_ASYNC_BLOCK(&ep->worker->async);
+            ucp_ep_set_failed(ep, UCP_NULL_LANE, UCS_ERR_CONNECTION_RESET);
+            UCS_ASYNC_UNBLOCK(&ep->worker->async);
+        }
```

Comment:
maybe kill receiver to generate the failure?
in general better use less UCP/UCT internals in tests and do it on API level as much as possible

Reply:
ok